Refactor how the Global Styles access and sets data

youknowriad opened the pull request:
This PR is not working yet but the code is in a state good enough to showcase what I'm trying to achieve:

 - Centralize all Global Styles Data Access and Write in a `global-styles/hooks` file with no dependency to context or whatsoever. Ultimately we should be able to move the whole `global-styles` folder anywhere.
 - Unify the way we retrieve settings, styles and update them, all of the following are correct:
 
 ```
 const [ color, setColor ] = useStyle( 'color.text' );
 const [ color, setColor ] = useStyle( 'color.text', 'core/paragraph' );
 const [ color, setColor ] = useStyle( 'color.text', 'core/paragraph', 'user' );
 const [ color, setColor ] = useStyle( 'color.text', 'core/paragraph', 'base' );
 
 const [ colorPalette, setColorPalette ] = useStyle( 'color.palette' );
 const [ colorPalette, setColorPalette ] = useStyle( 'color.palette', 'core/paragraph' );
 const [ colorPalette, setColorPalette ] = useStyle( 'color.palette', 'core/paragraph', 'user' );
 const [ colorPalette, setColorPalette ] = useStyle( 'color.palette', 'core/paragraph', 'base' );
 ```
 
 - The different components to render the different panels only need to know the `contextName` (blockName or undefined for root)

**Notes**

 - Right now I think this is not working yet because `useStyle` need to make transformations to values (it can't return the stored value). The current methods to do that in `editor/utils` look a bit too complex to me. I could use some help there cc @jorgefilipecosta. Can we simplify these? should we move them to `global-styles/utils`

youknowriad committed:
Refactor how the Global Styles access and sets data

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,210 @@
+function useGlobalStylesUserConfig() {
+	const globalStylesId = useSelect( ( select ) => {
+		return select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesUserEntityId;
```
Right now this is how we retrieve the current global styles entity ID. This is not great, we shouldn't be depending on the a setting of editSite store, The proposed endpoint here would solve this https://github.com/WordPress/gutenberg/discussions/35141

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,210 @@
+		'wp_global_styles',
+		'content',
+		globalStylesId
+	);
```
The way we retrieve user global styles (CPT endpoint) and base global styles (__experimentalGlobalStylesBaseStyles setting) is inconsistent, we should try to unify this. The proposed endpoint here would solve this #35141

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,210 @@
+	} );
+
+	return supportKeys;
+}
```
The new hooks useSetting and useStyle might be a bit less performant than the previous APIs in. `global-styles-provider` (which we should try to remove) because I'm not using "context" at all. I believe it's something we can improve later if we do notice issues, right now I wanted to focus on DevX

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/border-panel.js:
```diff
@@ -16,80 +16,90 @@ import { __ } from '@wordpress/i18n';
-		useHasBorderWidthControl( { supports, name } ),
+		useHasBorderColorControl( name ),
+		useHasBorderRadiusControl( name ),
+		useHasBorderStyleControl( name ),
```
Removing the supports property is a simple change that makes the code simpler 👍

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,210 @@
+		: `styles.blocks.${ blockName }.${ path }`;
+
+	const setStyle = ( newValue ) => {
+		const newUserConfig = cloneDeep( userConfig );
```
I guess we could easily improve the performance and even code readability by using the immutableSet function that we already use on other parts of the codebase:
```
function immutableSet( object, path, value ) {
	return setWith( object ? clone( object ) : {}, path, value, clone );
}
```

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/border-panel.js:
```diff
@@ -16,80 +16,90 @@ import { __ } from '@wordpress/i18n';
+	const supports = getSupportedGlobalStylesPanels( name );
 	return (
-		useSetting( 'border.customColor', name ) &&
+		useSetting( 'border.customColor', name )[ 0 ] &&
```
Having to use [ 0 ] when we just need to retrieve a value is the main downside of the new API, but I guess we will need to live with that.

ciampo replied:
To improve readability, maybe we could switch to a slightly more verbose syntax?

```js
// ...
const [ isEnabledInSettings ] = useSetting( 'border.customColor', name );
return isEnabledInSettings && supports.includes( 'borderColor' );
```

We'd then apply this syntax across all code changes in the PR

jorgefilipecosta replied:
Yes although more verbose, that syntax is more clear.

jorgefilipecosta commented:
I wonder a bit about performance and how things will scale but the current solution is not perfect either. I think if we follow the endpoint approach and remove the need to serialize and unserialize things become a little better. If we don't use the endpoint we may use the transient mechanism on entities.

ciampo commented on packages/edit-site/src/components/global-styles/screen-block.js:
```diff
@@ -6,21 +6,16 @@ import { getBlockType } from '@wordpress/blocks';
-				parentMenu={ '/blocks/' + name }
-				context={ blocks[ name ] }
-			/>
+			<ContextMenu parentMenu={ '/blocks/' + name } name={ name } />
```
Why is `ContextMenu` named this way? Maybe we can find a name that explains a bit better what's in it?

youknowriad replied:
Yes, I struggled to find a good name for this component. (It's the global styles menu for a give context, a context can be "root" or a specific block, potentially other kind of context in the future)

ciampo replied:
Got it! Maybe calling it `GlobalStylesMenu` would be a more clear name? Not big deal anyway.

ciampo commented on packages/edit-site/src/components/global-styles/border-panel.js:
```diff
@@ -16,80 +16,90 @@ import { __ } from '@wordpress/i18n';
  * Internal dependencies
  */
-import { useSetting } from '../editor/utils';
+import { getSupportedGlobalStylesPanels, useSetting, useStyle } from './hooks';
```
Hopefully not a silly question, but — is the old `useSetting` from `editor/utils` going to be replaced entirely by the new one in `hooks` (and removed) ?

And similarly, are the `getSetting`/`setSetting`/`getStyle`/`setStyle` functions from `global-styles-provider` also going to be replaced entirely from the new ones in `hooks`?

jorgefilipecosta replied:
Yes, that sounds to be like the plan to replace the context provider and the getters and setters with hooks.

youknowriad replied:
Yes, that's the idea.

youknowriad replied:
I removed a lot of code already, there might be other code that can be removed (maybe some constants in the wordpress/blocks package)

jorgefilipecosta commented:
The API is nice and I think provides a better development experience. The fact that we are not using context may bring some challenges. I guess with the plans we have for themes where during theme switching we may alternate between styles of one and styles of the other we may have cases where we want to preview the styles of different themes. With context that is simple it is matter of wrapping the preview contexts with different global style providers I guess with hooks we can also wrap with different EntityProviders? I would love to leave the door open for this use cases.

jorgefilipecosta committed:
Wire hooks to the variable engine

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/color-panel.js:
```diff
@@ -20,59 +21,67 @@ export function useHasColorPanel( { supports } ) {
 		( solids.length > 0 || areCustomSolidsEnabled );
 	const hasBackgroundColor =
 		supports.includes( 'backgroundColor' ) &&
-		isBackgroundEnabled &&
```
I removed these checks because they were making the color panel unavailable when it should be available? I guess we have a bug somewhere?

youknowriad replied:
I also noticed that the "custom color" links are not visible. I guess the `useSetting` calls are not working properly for some reason, I'll try to find out why.

youknowriad replied:
Solved in c8296e3

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/color-panel.js:
```diff
@@ -82,12 +91,9 @@ export default function ColorPanel( {
 			gradientValue: gradient,
-			onGradientChange: ( value ) =>
-				setStyle( name, 'background', value ),
+			onGradientChange: setGradient,
```
Currently setting a background color or gradient is not working. That happens because for example when we set the background color, the panel calls setBackgroundColor followed by setGradient to clear a gradient that may exist. The problem is that both setBackgroundColor and setGradient calls operate on the same object so setGradient reverts the change made by setBackgroundColor. setGradient should operate on the structure changed by setBackgroundColor. We have a race condition happening on our hooks.

youknowriad replied:
I had the same issue at some point in the `color.js` block support, We'll see how I fixed that.

youknowriad replied:
should be fixed in 8858351

jorgefilipecosta reviewed (commented):
I wired the hooks to our variable mechanism, excluding backgrounds and gradients (that have a specific issue), everything seems to be working, and we use preset variables, etc.
But I think what we have now is not something we can follow. We are constantly on each hook usage serializing and unserializing the structure, doing merge operations between the base and user styles, and processing the user styles to add and remove their origin on presets.
We need to hold a structure somewhere that represents the result of the merge between base and user styles, that structure is important for the variable retrieving mechanism to work on (and that mechanism is called on every useStyle usage), and that structure can also be used to compile the styles.
Previously that structure was on the GlobalStylesProvider.
I see two possibilities.
The first is we keep a GlobalStylesProvider that contains the merge results. That provider is also responsible for using our entity mechanism to make changes. The useStyle, useSetting hooks would interact with the provider, our API would exactly be the same as on this PR, but the hooks would only work when wrapped in a provider. On that provider, we would keep a central object with the result of merging user and base styles that provider would also compile the stylesheet etc., similar to what we have now.
The other possibility is having the result of the merge between user and base as a transient object (similar to how blocks are transient) on the entity where we have the user styles. Given that the base styles are immutable, this approach may work well.

My changes are in commit [`5852786` (#35264)](https://github.com/WordPress/gutenberg/pull/35264/commits/58527867e5b28d6d85a3c72bb0bab95d8681e163). Feel free to discard or apply any change to this commit, The objective of the commit is to show what we need to do to connect the hooks with our variable engine. The changes on file packages/edit-site/src/components/editor/utils.js are a port from PR https://github.com/WordPress/gutenberg/pull/34178; because the functions there are better isolated, we should base future work on this version.

youknowriad commented:
> The fact that we are not using context may bring some challenges

Using the context is not out of the question, For now, I wanted to keep this as simple as possible and avoid the "Provider" so that's what I removed the context, but it should be easy to bring back if we ever want to do it (without any change on the consumer components)

youknowriad commented:
> But I think what we have now is not something we can follow. We are constantly on each hook usage serializing and unserializing the structure, doing merge operations between the base and user styles, and processing the user styles to add and remove their origin on presets.

Yes, I'm aware of that, but did you notice any real performance issues? My initial plan was to optimize for devx and iterate to improve performance... Can we have a metric to track maybe?

youknowriad commented:
I think if we noticed perf issues... the context solution is probably the simplest one to go for.

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -39,6 +79,16 @@ function useGlobalStylesUserConfig() {
 			parsedConfig = content ? JSON.parse( content ) : {};
+			// It is very important to verify if the flag isGlobalStylesUserThemeJSON is true.
+			// If it is not true the content was not escaped and is not safe.
+			if ( ! parsedConfig.isGlobalStylesUserThemeJSON ) {
```
I don't understand this? How can we retrieve an unsafe content at this point? Should we escape in the backend instead?

jorgefilipecosta replied:
The WordPress escape filters don't provide information about which post type we are escaping etc, they just pass content to the filter, so we rely on the flag isGlobalStylesUserThemeJSON being present to know if we should escape the json or not. An attacker can easily bypass our escaping mechanism by not including the flag isGlobalStylesUserThemeJSON so we need on every usage to verify if the flag is present otherwise things may be unsafe.
If we have a specific endpoint that verification should be done there right now we don't have one so we need to keep this check on the client.

youknowriad replied:
It's still unclear to me, why not always escape?

youknowriad replied:
> If we have a specific endpoint that verification should be done there right now we don't have one so we need to keep this check on the client.

I guess that's the missing piece for me, let's try to bring that endpoint after this PR. it seems it would solve a lot of complexity/issues.

jorgefilipecosta replied:
> It's still unclear to me, why not always escape?

Because you just receive a string of content, you can not escape a given JSON or HTML or anything in the post content with the theme.json rules. That content may be a JSON from a plugin following totally different sets of rules for what should be escaped. The escape functions does not receive anything besides a string of content that flag indicates that content is a user theme.json.

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -105,10 +163,10 @@ export function useSetting( path, blockName, source = 'all' ) {
 	switch ( source ) {
 		case 'all':
-			result = get( userConfig, finalPath ) ?? getBaseSetting();
+			result = get( userConfig, finalPath, {} ).user ?? getBaseSetting();
```
This change is confusing to me. Why do we have a `.user` subkey in the configs?

youknowriad replied:
Also, this change makes it so that `all` and `user` are basically similar and we don't want that: `all` is meant to returned the merged configs, while `user` is meant to return only the user changes (excluding the base config)

youknowriad replied:
This seems to be related to having `.user` as values for some configs, now that we can easily access all "user", "base" values directly maybe we can normalize all configs and drop this sunkey entirely.

jorgefilipecosta replied:
I think we still need the subkey, base includes core and theme so we need the subkey to differentiate e.g: the UI will show both. So I guess we can normalize to always have subkeys as it is doing now.

youknowriad replied:
> the UI will show 

The UI can do

```
useSetting( 'something', context, 'all' ); // merged
useSetting( 'something', context, 'user' ); // user
useSetting( 'something', context, 'base' ); // base
``` 

so the object itself shouldn't have the keys. it's just adds complexity and it's not consistent across settings and styles (and even inside settings)

jorgefilipecosta replied:
How would the UI differentiate between core and theme palettes? The base needs to have subjects for core and theme, the UI shows them differently.

youknowriad replied:
It could be a different API call (query arg or something). Anyway, it feels like implementation detail for now but I feel it would simplify a lot of things if the "configs" had a consistent shape, regardless how they're retrieved and the arguments (merged, user, base,...).

jorgefilipecosta replied:
> it would simplify a lot of things if the "configs" had a consistent shape, regardless of how they're retrieved and the


I agree with that but I think we are being consistent if all the presets no matter where they come from are grouped by origin e.g: in base object we have { core: [], theme: [] }, in user we have { user: [] }, in the merged object we have { core: [], theme: [], user: [] }.

Grouping by origin we have a consistent shape where we can always work on.

> This change is confusing to me. Why do we have a .user subkey in the configs?

The addition of the user subkey is done in order to improve consistency we make presets always have an origin subkey.

But I think I understand the point that in an user object having a user subkey is redundant it is also ok for me if the user object does not have the subkey as long as the base ( { core: [], theme: [] } ) and merged objects ( { core: [], theme: [], user: [] } ) have them.

youknowriad replied:
> But I think I understand the point that in an user object having a user subkey is redundant it is also ok for me if the user object

Yes, this is the issue I have with this but it's not specific to the user object for me, it's for all objects. We should have a "core object",  a "theme object", a "user object" and a "merged object" all with no subkeys.

jorgefilipecosta replied:
> We should have a "core object", a "theme object", a "user object" and a "merged object" all with no subkeys.

On the block editor, we are also going to show different color palettes per origin. So we thought it would be better to pass a single object to the block editor where presets are keyed by origin instead of passing multiple objects and that by origin subkey approach appeared. We can have multiple objects but in that case, we also need to make the block editor aware of the multiple objects, the block editor will never need anything from other sources besides the presets so the by origin approach seemed a better fit.

youknowriad replied:
Got it 👍 I guess there's drawbacks for each, let's keep it as is then and continue monitoring for the future.

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -19,9 +19,49 @@ import {
+	PRESET_METADATA,
+	getValueFromVariable,
+	getPresetVariableFromValue,
+} from '../editor/utils';
```
Can we move these utils file to the`global-styles` folder ?

jorgefilipecosta replied:
Yes, we can move it there for now, but ideally, it should be part of the block editor as I'm proposing on https://github.com/WordPress/gutenberg/pull/34178.

youknowriad committed:
Add missing return

youknowriad committed:
Make global styles updates independent of the component lifecycle

youknowriad committed:
Simplify global styles renderer and remove useless code

youknowriad committed:
Remove old useSetting

youknowriad commented:
@jorgefilipecosta I think some of your changes might have broken the "blocks" panels.

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,322 @@
+		case 'user':
+			result = getSettingValue( userConfig );
+			break;
+		case 'base':
```
Related to another discussion point, I think we can not have only base there are cases where we need to differentiate between 'core', and 'theme', e.g: on the palette we show the default (core), the theme and the user one.

youknowriad replied:
Right now, there's no way to retrieve these separately (core and theme), once we have that in the endpoint we should add these.

oandregal commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,322 @@
+function useGlobalStylesBaseConfig() {
+	const baseConfig = useSelect( ( select ) => {
+		return select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesBaseStyles;
```
Working on other things I've noticed that we're dumping a lot of info we don't need at `__experimentalGlobalStylesBaseStyles`. It should only contain the base styles but not the settings and other properties of the `theme.json`.

I believe we may have done this because we also needed the base presets. Note that we have now available the presets for each origin (core, theme, user) under `__experimentalFeatures`, so `__experimentalGlobalStylesBaseStyles` can go back to styles-only.

youknowriad replied:
My idea at the moment is that we should remove both of them in favor of a `core-data` entity endpoint.

oandregal replied:
Follow-up at https://github.com/WordPress/gutenberg/pull/35458 (getting rid of `settings` has proven convoluted, so I'm leaving that for later).

youknowriad committed:
Fix block global styles

youknowriad committed:
Move all the global styles logic to the global styles folder

youknowriad committed:
Add documentation

youknowriad committed:
Fix unit tests

youknowriad commented:
I think this is now ready, I've organized the folder properly, added some documentation and fixed the bugs I saw in the "blocks" panels.

It would be good to land this quickly to unblock further work. Can I get a review (I know it's hard because it's a big PR but it doesn't have to be perfect, this is experimental UI still)

ciampo commented:
I will leave the overall _code_ review to @oandregal and @jorgefilipecosta , since I'm not very experienced in this area of the repo and its general architecture.

But I did some smoke testing and came across a few issues:

I wasn't able to add a custom color to the palette

https://user-images.githubusercontent.com/1083581/135995197-44393ee0-7ed8-46a0-8a6d-9e303359b509.mp4

I wasn't able to change the text color of the `Heading` block

https://user-images.githubusercontent.com/1083581/135995184-e79700f9-426c-459a-b780-5671f817594b.mp4


I haven't smoke tested every single combination of block/setting, so there may be more issue like this that I haven't found yet.

youknowriad committed:
Fix settings updates

youknowriad commented:
@ciampo Both the issues are fixed for me, for the second one I'm not sure if it's a real issue or something about style specificity battle in your particular example.

ciampo commented:
> Both the issues are fixed for me, for the second one I'm not sure if it's a real issue or something about style specificity battle in your particular example.

I confirm that the first issue is fixed.

Regarding the second issue (setting the text color of the Heading block), you were correct on it not being an issue. I believe the "conflict" comes from the fact that in my example, the headings were also links. I tried to create a separate heading without any linked text, and everything worked as expected

oandregal commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,347 @@
+	return [ resultWithFallback, setSetting ];
+}
+
+export function useStyle( path, blockName, source = 'all' ) {
```
What are your thoughts on accommodating elements in this API? I'm thinking about this use case: UI controls for elements.

It is something we'll add to the global styles sidebar soon. Potentially, we should also add it to the blocks UI panels (both in the block sidebar & the global sidebar): for example, at the moment, users can't update all the styles of link color despite the theme can. To support this, we'll need the ability to address elements individually. Elements can be at the top or within a block.

In [this other conversation](https://github.com/WordPress/gutenberg/pull/34843#issuecomment-920030977) I was thinking along these lines:

`useSetting( path, { block: 'blockName', element: 'elementName' }, source )`

- top level: `useStyle( 'group.property' )`
- block level: `useStyle( 'group.property', { block: 'block/name' } )`
- element at top-level: `useStyle( 'group.property', { element: 'element/name' } )`
- element at block-level: `useStyle( 'group.property', { block: 'block/name', element: 'element/name' } )`

Whatever the choice to address this, I feel it should be the same in both the client and the server

oandregal replied:
I'd probably do the same for `useSetting`, although at the moment elements can't be in the settings. They may need that when/if they get added to the block UI panels. We may also have other contexts than blocks, so being an object will allow us to grow the API easily.

youknowriad replied:
Well right now you can do `useStyle( 'elements.link.background' )` if you want to get/set the background of a link for instance. I agree it's not perfect yet and thinking of "elements" as "contexts" might make sense. These APIs are internal for now so there's room for iterations to see what would work well or not.

oandregal replied:
Yeah, I know. This PR shouldn't be blocked by this. it's something we can iterate on. Though my concern with that approach is that it ties consumers to a specific `theme.json` shape, which we tried to avoid by separating "path" and "context" in the hooks/utilities.

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -0,0 +1,347 @@
+	];
+};
+
+export function useSetting( path, blockName, source = 'all' ) {
```
The useSetting hook on the block editor has a slightly different API the source is part of the path.
useSetting( 'color.palette' ) returns merged color palette.
useSetting( 'color.palette.user' ) returns user color palette.

We may need to consolidate these two approaches.

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/color-panel.js:
```diff
@@ -52,27 +51,44 @@ export default function ColorPanel( {
+		name
+	);
+	const [ userLinkColor ] = useStyle(
+		'elements.link.color.text',
```
I'm not sure if elements should be treated as part of the path or if they should be something like a block in their own right. But I guess we can follow this approach for now.

youknowriad replied:
Agreed, it's not clear. See https://github.com/WordPress/gutenberg/pull/35264#discussion_r723280161

oandregal commented on packages/edit-site/src/components/global-styles/typography-panel.js:
```diff
@@ -26,14 +27,16 @@ export function useHasTypographyPanel( { supports, name } ) {
+function useHasLineHeightControl( name ) {
+	const supports = getSupportedGlobalStylesPanels( name );
 	return (
 		useSetting( 'typography.customLineHeight', name ) &&
```
All these `useSetting` calls need to pull the value (either `useSetting( 'typography...', name)[ 0 ]` or `[ value ] = useSetting('typography...', name )`).

oandregal commented on packages/edit-site/src/components/global-styles/dimensions-panel.js:
```diff
@@ -14,31 +14,34 @@ import { __experimentalUseCustomSides as useCustomSides } from '@wordpress/block
-function useHasPadding( { name, supports } ) {
+function useHasPadding( name ) {
+	const supports = getSupportedGlobalStylesPanels( name );
 	const settings = useSetting( 'spacing.customPadding', name );
```
All these `useSetting` calls need to pull the value (either `useSetting( 'typography...', name)[ 0 ]` or `[ value ] = useSetting('typography...', name )`), correct?

youknowriad committed:
Fix useSetting calls

jorgefilipecosta reviewed (commented):
Hi @youknowriad,
I think the performance issue is noticeable even when no changes are applied and just by navigating around the panel.

I did this test in chrome with 6x CPU throttling ( I think many people have CPU even lower than my machine with throttling). The time it takes to open the color section is huge.

In this PR:

https://user-images.githubusercontent.com/11271197/136236899-b1bc2ff4-95f9-4cea-8c59-f74097b943fb.mp4

In trunk:

https://user-images.githubusercontent.com/11271197/136237032-c096f2c3-fa6d-4dfb-a67f-729c808da02e.mp4

oandregal commented on packages/edit-site/src/components/global-styles/color-palette-panel.js:
```diff
@@ -27,49 +21,25 @@ import { store as editSiteStore } from '../../store';
-	);
+export default function ColorPalettePanel( { name } ) {
+	const [ colors, setColors ] = useSetting( 'color.palette', name );
+	const userColors = useSetting( 'color.palette', name, 'user' );
```
Same as with typography and dimension panels: for these instructions, we need to pull the value from the array returned by `useSetting`.

jorgefilipecosta commented:
I also notice some lag when moving the color picker around even without throttling (for example for a custom background color), the UI feels slow and takes some time to represent the choose background color. The trunk is not perfect on this either.

oandregal commented on packages/edit-site/src/components/global-styles/screen-block-list.js:
```diff
@@ -38,15 +35,10 @@ function BlockMenuItem( { block } ) {
 		<>
 			<ScreenHeader back="/" title={ __( 'Blocks' ) } />
-			{ visibleBlocks.map( ( block ) => (
+			{ getBlockTypes().map( ( block ) => (
```
Reviewing this change, I've noticed that `hasBlockMenuItem` a few lines above is missing the `hasBorderPanel || hasDimensionsPanel` checks as well, so we won't list blocks that support only border or dimensions (or both).

youknowriad replied:
It's included in `hasLayoutPanel` so we should be fine I think.

oandregal replied:
Oh, I missed that `hasLayoutPanel` is precisely the two I brought up. It's all good.

jorgefilipecosta reviewed (commented):
When I apply a link color at the block level(paragraph link color) the color is not reflected on the editor but is reflected on the frontend. Inspecting the styles on the editor I see the style is not present. On trunk paragraph link color appears as expected.

jorgefilipecosta reviewed (approved):
Thank you for all the work on this refactor 👍 I noticed an issue with the block level link color. I also think we should follow up and improve the performance with some throttling the slowness is noticeable just by moving between panels, and even without throttling, I notice some slowness while changing styles fast.
I think these are things we can work on after, provided the link color regression is solved I think we can merge this PR.

youknowriad committed:
Fix elements outputs

youknowriad committed:
Fix wrong useSetting usage

youknowriad commented:
@jorgefilipecosta I fixed the issue 👍 bad refactoring :) 

For the performance side, is this something you want to try separately? (Maybe move the `useGlobalStylesConfig` call to a context provider). Also would be good if we can come up with some metric to track in the site editor perf tests.

jorgefilipecosta commented:
> For the performance side, is this something you want to try separately? 

yes it is something to look separately and not a blocker.

youknowriad committed:
Fix unit tests
